File: src/main/java/ms/aurora/loader2/ClientLoader.scala
package ms.aurora.loader2

import java.applet.Applet
import scala.beans.BeanProperty
import org.apache.log4j.Logger
import java.net.{JarURLConnection, URL}
import ms.aurora.transform.ClientClassLoader
import ms.aurora.sdn.net.api.Hooks.getHooks

/**
 * Applet loader responsible for loading the client
 * based on a ClientConfiguration.
 * @author rvbiljouw
 */
class ClientLoader {
  val logger = Logger.getLogger(classOf[ClientLoader])
  @BeanProperty var applet: Applet = null

  /**
   * Loads the game applet based on the information
   * obtained from the client configuration.
   * @param configuration Client configuration
   */
  def start(configuration: ClientConfig) {
    if (!isLoaded) {
      logger.info("Starting new applet")
      val stub = new ClientStub(configuration)
      prepareApplet(configuration)
      applet.setStub(stub)
      applet.init()
      applet.start()
    } else {
      logger.warn("Applet wasn't started, call stop()!")
    }
  }

  /**
   * Stops the applet
   */
  def stop() {
    applet.stop()
    applet.destroy()
    applet = null
    System.gc()
  }

  /**
   * Checks if the applet is loaded and active
   * @return true when loaded and active
   */
  def isLoaded: Boolean = {
    applet != null && applet.isActive
  }

  /**
   * Establishes a JarURLConnection based on information
   * from the client configuration
   * @param config Client configuration
   * @return a jar URL connection
   */
  private def establishJarConnection(config: ClientConfig): JarURLConnection = {
    val docBase = config.getDocumentBase
    val archive = config.getArchiveName
    new URL("jar:" + docBase + "/" + archive + "!/")
      .openConnection().asInstanceOf[JarURLConnection]
  }

  /**
   * Prepares the applet for loading.
   * First it creates a transforming class loader for
   * the client jar, and then it requests the main class and
   * instantiates it. The applet is stored into the applet var
   * @param config Client configuration
   */
  private def prepareApplet(config: ClientConfig) {
    val connection = establishJarConnection(config)
    val loader = new ClientClassLoader(getHooks, connection.getJarFile)
    val mainClass = loader.loadClass(config.getMainClass)
    applet = mainClass.newInstance.asInstanceOf[Applet]
  }

}

File: src/main/java/ms/aurora/loader/ClientLoader.java
package ms.aurora.loader;

import ms.aurora.loader.impl.ClientStub;
import ms.aurora.sdn.net.api.Hooks;
import ms.aurora.transform.ClientClassLoader;
import org.apache.log4j.Logger;

import java.applet.Applet;
import java.io.IOException;
import java.net.JarURLConnection;
import java.net.URL;

/**
 * @author rvbiljouw
 */
public class ClientLoader {
    private static final Logger logger = Logger.getLogger(ClientLoader.class);
    private final ClientConfig config;
    private Applet applet;

    public ClientLoader(ClientConfig config) {
        this.config = config;
    }

    public void start() {
        if (!isLoaded()) {
            try {
                logger.info("Starting new applet");
                ClientStub stub = new ClientStub(config);
                prepareApplet();
                applet.setStub(stub);
                applet.init();
                applet.start();
            } catch (ReflectiveOperationException e) {
                logger.error("Initialization failed", e);
            } catch (IOException e) {
                logger.error("I/O fault", e);
            }
        } else {
            logger.warn("Applet waswn't started, call stop()!");
        }
    }

    public void stop() {
        if (applet != null) {
            applet.stop();
            applet.destroy();
            applet = null;
        }
        System.gc();
    }

    public boolean isLoaded() {
        return applet != null && applet.isActive();
    }

    private JarURLConnection establishJarConnection() throws IOException {
        String docBase = config.getDocumentBase();
        String archive = config.getArchiveName();
        return (JarURLConnection)
                new URL("jar:" + docBase + archive + "!/")
                        .openConnection();
    }

    private void prepareApplet() throws ReflectiveOperationException, IOException {
        JarURLConnection connection = establishJarConnection();
        ClientClassLoader loader = new ClientClassLoader(
                Hooks.getHooks(), connection.getJarFile());
        Class<?> mainClass = loader.loadClass(config.getMainClass());
        applet = (Applet) mainClass.newInstance();
    }

    public Applet getApplet() {
        return applet;
    }

    public void setApplet(Applet applet) {
        this.applet = applet;
    }
}
